File: cpp/299_BullsAndCows.cpp
class Solution {
public:
    string getHint(string secret, string guess) {
        unordered_map<char, int> secret_counts;
        for (char c : secret)
            secret_counts[c] = secret_counts.find(c) == secret_counts.end() ? 1 : secret_counts[c] + 1;
        
        int bulls = 0, cows = 0;
        for (int i = 0; i < secret.length(); ++i) {
            char c = secret[i];
            if (c == guess[i]) {
                ++bulls;
                --secret_counts[c];
            }
        }
        for (int i = 0; i < guess.length(); ++i) {
            char c = guess[i];
            if (secret_counts.find(c) != secret_counts.end() && 
                secret_counts[c] > 0 &&
                c != secret[i]) {
                ++cows;
                --secret_counts[c];
            }
        }
        
        return to_string(bulls) + "A" + to_string(cows) + "B";
    }
};

File: java/299_BullsAndCows.java
class Solution {
    public String getHint(String secret, String guess) {
        HashMap<Character, Integer> counts = new HashMap<Character, Integer>();
        int bulls = 0, cows = 0;
        for (int i = 0; i < secret.length(); ++i) {
            char s = secret.charAt(i), g = guess.charAt(i);
            if (s == g) {
                ++bulls;
            } else {
                if (counts.getOrDefault(s, 0) < 0) ++cows;
                if (counts.getOrDefault(g, 0) > 0) ++cows;
                
                // secret and guess keys modify distance from 0 vals
                counts.put(s, counts.getOrDefault(s, 0) + 1);
                counts.put(g, counts.getOrDefault(g, 0) - 1);
            }
        }
        return bulls + "A" + cows + "B";
    }
}
